client/src/portal/mod.rs

```diff
@@ -374,12 +374,76 @@ impl Keyring {
     }
 
     /// Return key, derive and store it first if not initialized
-    async fn derive_key(&self) -> &Key {
-        if self.key.get().is_none() {
-            let key = self.keyring.read().await.derive_key(&self.secret);
-            self.key.set(key).unwrap();
+    async fn derive_key(&self) -> Arc<Key> {
+        let keyring = Arc::clone(&self.keyring);
+        let secret = Arc::clone(&self.secret);
+
+        let mut key_lock = self.key.lock().await;
+        if key_lock.is_none() {
+            #[cfg(feature = "async-std")]
+            let key = blocking::unblock(move || {
+                async_io::block_on(async { keyring.read().await.derive_key(&secret) })
+            })
+            .await;
+            #[cfg(feature = "tokio")]
+            let key =
+                tokio::task::spawn_blocking(move || keyring.blocking_read().derive_key(&secret))
+                    .await
+                    .unwrap();
+
+            *key_lock = Some(Arc::new(key));
         }
-        self.key.get().unwrap()
+
+        Arc::clone(key_lock.as_ref().unwrap())
+    }
+
+    /// Change keyring secret
+    ///
+    /// # Arguments
+    ///
+    /// * `secret` - The new secret to store.
+    pub async fn change_secret(&mut self, secret: Secret) -> Result<(), Error> {
+        #[cfg(feature = "tracing")]
+        tracing::debug!("Changing keyring secret and key");
+
+        let keyring = self.keyring.read().await;
+        let key = self.derive_key().await;
+
+        let mut items = Vec::new();
+        for item in &keyring.items {
+            items.push(item.clone().decrypt(&key)?);
+        }
+        drop(keyring);
+
+        self.secret = Arc::new(secret);
+
+        let keyring = Arc::clone(&self.keyring);
+        let secret = Arc::clone(&self.secret);
+        #[cfg(feature = "async-std")]
+        let key = blocking::unblock(move || {
+            async_io::block_on(async { keyring.read().await.derive_key(&secret) })
+        })
+        .await;
+        #[cfg(feature = "tokio")]
+        let key = tokio::task::spawn_blocking(move || keyring.blocking_read().derive_key(&secret))
+            .await
+            .unwrap();
+
+        let mut key_lock = self.key.lock().await;
+        *key_lock = Some(Arc::new(key));
+        let key = Arc::clone(key_lock.as_ref().unwrap());
+        drop(key_lock);
+
```

Comment:
all this code is what derive_key does, not sure why you want to duplicate it here

Reply:
yeah, it is duplicate code :)

`derive_key(&self)` only sets the `key` if it's value is [not set](https://github.com/bilelmoussaoui/oo7/blob/dc77973722536a0f985339d9982fe3f5ab0a03c0/client/src/portal/mod.rs#L382), right. since we're changing the Keyring secret here at `change_secret`, we need to derive a key from the new secret and update `Keyring`'s `secret`, `key` fields. otherwise `change_secret` doesn't  work and we get `MacError`.

Reply:
I still don't understand the reasoning here, you can unset the key before calling derive_key.

Reply:
yes! that way is much cleaner. done, thanks!